### ISSUE TITLE
Add rules to enforce custom properties for types

### ISSUE THREAD

**Issue by silverwind** [stylelint/stylelint#6368]
### What is the problem you're trying to solve?

I would like a rule that allows colors only within custom properties, e.g. any color use outside a custom property should raise an error. This ensures that all colors are overridable via custom properties which enables color theming.

### What solution would you like to see?

A new rule `color-require-custom-property`.

Fail:
```css
a {
  color: red; /* Error: Color 'red' must be defined as a custom property. */
}
```

Pass:
```css
:root {
  --color-link: red;
}

a {
  color: var(--color-link);
}
```

Maybe the rule could even be made more generic to cover more fundamental CSS types, e.g.

````yaml
custom-property-require: error, [color, length]

**Comment by jeddy3** [comment]
@silverwind Thanks for the request and for using the template.

You can use the [`stylelint-declaration-strict-value`](https://github.com/AndyOGo/stylelint-declaration-strict-value) plugin to do this. It has [limitations around shorthand properties](https://github.com/AndyOGo/stylelint-declaration-strict-value#css-shorthand-syntax), though. 

The [`<color>` type is used in lots of places](https://csstree.github.io/docs/syntax/#Type:color).

Alternatively, you can use a combination of `no` and `disallowed-list` built-in rules:

- [color-named](https://stylelint.io/user-guide/rules/list/color-named/)
- [color-no-hex](https://stylelint.io/user-guide/rules/list/color-no-hex/)
- [function-disallowed-list](https://stylelint.io/user-guide/rules/list/function-disallowed-list/)

```js
// .stylelintrc.json
{
  "rules": {
    "color-named": "never",
    "color-no-hex": true,
    "function-disallowed-list": ["rgb", "hsl", ..]
  }
}
```

```css
/* stylelint-disable color-named */
:root {
  --color-link: red;
  ..
}
/* stylelint-enable color-named */

a {
  color: var(--color-link);
}

..
```

This approach doesn't have a shorthand property limitation. It only works for colours, though.

I think we can only add built-in rules that enforce custom properties for specific _types_, rather than _properties_, if we use a finer-grained parser for these rules. I'll label as discussion in case anyone else has any thoughts on this. I haven't looked at https://github.com/stylelint/stylelint/issues/5586 recently, but perhaps SWC or the parser behind [Lightning CSS](https://lightningcss.dev/) offer this granularity.

In the meantime, you can choose either of the approaches above or [crafting a plugin](https://stylelint.io/developer-guide/plugins) for your specific needs.

**Comment by silverwind** [comment]
I was able to make `stylelint-declaration-strict-value` work with a bit of configuration, so I'll close this. Thanks again for pointing me in the right direction.
